### PR TITLE
added checks for public services on compiler passes that use service ids and not references

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php [hunk 1]
```diff
@@ -231,6 +231,7 @@ public function testRuntimeLoader()
 
         $loader = $container->getDefinition('twig.runtime_loader');
         $args = $loader->getArgument(1);
+print_r($loader->getArguments());
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: This looks like a debug leftover

Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: oops

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php [hunk 2]
```diff
@@ -28,7 +28,17 @@ public function process(ContainerBuilder $container)
...
+
+            if (!$definition->isPublic()) {
+                throw new InvalidArgumentException(sprintf('The service "%s" must be public as it can be lazy-loaded.', $id));
+            }
+
+            if ($definition->isAbstract()) {
+                throw new InvalidArgumentException(sprintf('The service "%s" must not be abstract as it can be lazy-loaded.', $id));
+            }
+
+            $validators[$definition->getClass()] = $id;
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: shouldn't this be done in older branches ?

Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: see #20116

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/RuntimeLoaderPass.php [hunk 3]
```diff
@@ -28,7 +28,17 @@ public function process(ContainerBuilder $container)
...
         foreach ($container->findTaggedServiceIds('twig.runtime') as $id => $attributes) {
-            $mapping[$container->getDefinition($id)->getClass()] = $id;
+            $def = $container->getDefinition($id);
+
+            if (!$def->isPublic()) {
+                throw new InvalidArgumentException(sprintf('The service "%s" must be public as it can be lazy-loaded.', $id));
+            }
+
+            if ($def->isAbstract()) {
+                throw new InvalidArgumentException(sprintf('The service "%s" must not be abstract as it can be lazy-loaded.', $id));
```
Comment (lyrixx) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/RuntimeLoaderPass.php]: The message here is not very accurate. An abstract service is not "get() able". It's not about lazy-loading.

**Pull request merged**